File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool three(vector<vector<char>> board, int index){
        int row = index / 3 * 3;
        int col = index % 3 * 3;
        unordered_map<int, int> mp;
        for(int i=0;i<3;i++){
            for(int j=0;j<3;j++){
                if(board[row+i][col+j] == '.') continue;
                mp[board[row+i][col+j]]++;
                if(mp[board[row+i][col+j]] > 1) return false;
            }
        }

        return true;
    }
    bool rowCol(vector<vector<char>> board, int row, int col, int target){
        for(int i=col+1;i<9;i++){
            if(board[row][i] == target) return false;
        }
        for(int i=row+1;i<9;i++){
            if(board[i][col] == target) return false;
        }
        return true;
    }
    bool isValidSudoku(vector<vector<char>>& board) {
        for(int i=0;i<9;i++){
            bool temp = three(board, i);
            if(!temp) return false;
        }
        for(int i=0;i<9;i++){
            for(int j=0;j<9;j++){
                if(board[i][j] == '.') continue;
                bool temp = rowCol(board, i, j, board[i][j]);
                if(!temp) return false;
            }
        }
        return true;
    }

};

File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        for(int i=0;i<9;i++){
            if(!isValidRow(board[i]) || 
                !isValidCol(board, i) || 
                !isValidBox(board, i)) return false;
        }
        return true;
    }

    private boolean isValidRow(char[] row){
        int[] arr = new int[10];
        for(int i=0;i<9;i++){
            if(row[i] == '.') continue;
            else if(arr[row[i]-'0'] > 0) return false;
            arr[row[i]-'0']++;
        }
        return true;
    }

    private boolean isValidCol(char[][] col, int index){
        int[] arr = new int[10];
        for(int i=0;i<9;i++){
            if(col[i][index] == '.') continue;
            else if(arr[col[i][index]-'0'] > 0) return false;
            arr[col[i][index]-'0']++;
        }
        return true;
    }


    private boolean isValidBox(char[][] box, int index){
        int[] arr = new int[10];
        int row = (index / 3) * 3;
        int col = (index % 3) * 3;
        for(int i=0;i<3;i++){
            for(int j=0;j<3;j++){
                if(box[row+i][col+j] == '.') continue;
                else if(arr[box[row+i][col+j]-'0'] > 0) return false;
                arr[box[row+i][col+j]-'0']++;
            }
        }
        return true;
    }
}